text/0000-dyno.md

```diff
@@ -0,0 +1,294 @@
+
+A user of `Error` trait objects can call `get_context_ref` to access data by type which might be carried by an `Error` object. The function returns an `Option` and will return `None` if the requested type of data is not carried. For example, to access a backtrace (in a world where the `Error::backtrace` method has been removed): `e.get_context_ref::<Backtrace>()` (specifying the type using the turbofish syntax may not be necessary if the type can be inferred from the context).
+
+Lets examine the changes to `Error` required to make this work:
+
+```rust
+pub mod error {
+    pub trait Error: Debug + Provider {
+        ...
+        fn provide_context<'a>(&'a self, _req: Requisition<'a, '_>) {}
+    }
+
+    impl<T: Error> Provider for T {
+        fn provide<'a>(&'a self, req: Requisition<'a, '_>) {
+            self.provide_context(req);
+        }
+    }
+
+    impl dyn Error {
+        ...
+        pub fn get_context_ref<T: ?Sized + 'static>(&self) -> Option<&T> {
+            crate::provide_any::request_by_type_tag::<'_, tags::Ref<T>>(self)
+        }
+    }
+}
+```
+
+`get_context_ref` is added as a method on `Error` trait objects (`Error` is also likely to support similar methods for values and possibly other types, but I will elide those details), it simply calls `provide_any::request_by_type_tag` (we'll discuss this function and the `tags::Ref` passed to it below). But where does the context data come from? If a concrete error type supports backtraces, then it must override the `provide_context` method when implementing `Error` (by default, the method does nothing, i.e., no data is provided, so `get_context_ref` will always returns `None`). `provide_context` is used in the blanket implementation of `Provider` for `Error` types, in other words `Provider::provide` is delegated to `Error::provide_context`.
+
+In `provide_context`, an error type provides access to data via a `Requisition` object, e.g., `req.provide_ref::<Backtrace>(&self.backtrace)`. The type of the reference passed to `provide_ref` is important here (and we encourage users to use explicit types with turbofish syntax even though it is not necessary). When a user calls `get_context_ref`, the requested type must match the type of an argument to `provide_ref`, e.g., the type of `&self.backtrace` is `&Backtrace`, so a call to `get_context_ref::<Backtrace>()` will return a reference to `self.backtrace`. An implementer can make multiple calls to `provide_ref` to provide multiple data with different types.
+
+Note that `Requisition` has methods for providing values as well as references, and for providing more complex types. These will be covered in the next section.
+
+The important parts of `provide_any` are
+
+```rust
+pub mod provide_any {
+    pub trait Provider {
+        fn provide<'a>(&'a self, req: Requisition<'a, '_>);
```

Comment:
I've currently got a bit of a mental model of `req.provide_ref` as analogous to making a field `pub`. The error use-case doesn't have a need for it, but I wonder if there's any consideration then for something like `provide_mut`, and whether the data you offer up through `provide` would be the same as `provide_mut`. In the `pub` line of thinking it seems like you would expect them to be the same (there's no `pub(const)` for instance).

I'm wondering mainly whether `provide_mut` is:

- A valid extension at all.
- One that would be reasonable as a backwards compatible addition with some empty default, assuming you might want to expose a different set of fields mutably (breaks the link between `pub`).

My mental model may also be wrong.

Reply:
> My mental model may also be wrong.

It probably is, on a second pass they're more like getters, so there doesn't have to be any symmetry between what you can get and what you can set.

Reply:
Implementers can return &mut refs if they want users to be able to modify things, so there is no need for an explicit provide_mut method